oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/worker/MetricsPersistentWorker.java

```diff
@@ -246,10 +258,35 @@ private void nextWorker(Metrics metrics) {
      * Load data from the storage, if {@link #enableDatabaseSession} == true, only load data when the id doesn't exist.
      */
     private void loadFromStorage(List<Metrics> metrics) {
+        final long currentTimeMillis = System.currentTimeMillis();
         try {
-            List<Metrics> notInCacheMetrics = metrics.stream()
-                                                     .filter(m -> !context.containsKey(m) || !enableDatabaseSession)
-                                                     .collect(Collectors.toList());
+            List<Metrics> notInCacheMetrics =
+                metrics.stream()
+                       .filter(m -> {
+                           final Metrics cachedValue = context.get(m);
+                           // Not cached or session disabled, the metric could be tagged `not in cache`.
+                           if (cachedValue == null || !enableDatabaseSession) {
+                               return true;
+                           }
+                           // The metric is in the cache, but still we have to check
+                           // whether the cache is expired due to TTL.
+                           // This is a cache-DB inconsistent case:
+                           // Metrics keep coming due to traffic, but the entity in the
+                           // database has been removed due to TTL.
+                           if (!model.isTimeRelativeID() && supportUpdate) {
+                               // Mostly all updatable metadata level metrics are required to do this check.
+                               final long metricsTimestamp = TimeBucket.getTimestamp(
+                                   cachedValue.getTimeBucket(), model.getDownsampling());
+                               // If the cached metric is older than the TTL indicated.
+                               if (currentTimeMillis - metricsTimestamp > metricsDataTTLInMillis) {
+                                   // The expired metrics should be tagged `not in cache` directly.
+                                   return true;
```

Comment:
I don't remove the existing cache, as in this case, the expired key should be overridden directly in the next `put`/`read` stage.

Reply:
This is ok for now, but when we support MAL (or OAL) hot reload dynamically, it’s possible users will remove some metrics and this cache will be leaked.

Reply:
I think it wouldn't. There are 2 reason
1. Session is built based on hot access. Once there is no traffic, the session gets removed after minutes.
2. Hot reload should remove worker instances too. The session is attached, so should be removed too.

Also, this PR is removing hot data, it will be in the cache one way or another. Such as reloading from database, due to merging.